<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Knowledge Book</title>
  <meta name="description" content="This book contains some knowledge I have adquired during my PhD studies." />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Knowledge Book" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book contains some knowledge I have adquired during my PhD studies." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Knowledge Book" />
  
  <meta name="twitter:description" content="This book contains some knowledge I have adquired during my PhD studies." />
  

<meta name="author" content="Sonia García-Ruiz (s.ruiz@ucl.ac.uk)" />


<meta name="date" content="2021-08-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="github.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Sonia's Knowledge Book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html"><i class="fa fa-check"></i><b>1</b> Analysis of Genetic Association Studies</a><ul>
<li class="chapter" data-level="1.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#useful-linux-commands"><i class="fa fa-check"></i><b>1.1</b> Useful Linux commands</a></li>
<li class="chapter" data-level="1.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#useful-r-commands"><i class="fa fa-check"></i><b>1.2</b> Useful R commands</a></li>
<li class="chapter" data-level="1.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#variant-calling-from-sequencing-data"><i class="fa fa-check"></i><b>1.3</b> Variant Calling from sequencing data</a><ul>
<li class="chapter" data-level="1.3.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#generate-genotype-calls-for-a-single-sample"><i class="fa fa-check"></i><b>1.3.1</b> Generate genotype calls for a single sample</a></li>
<li class="chapter" data-level="1.3.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#generating-genotypes-from-multiple-.bam-files"><i class="fa fa-check"></i><b>1.3.2</b> Generating genotypes from multiple .bam files</a></li>
<li class="chapter" data-level="1.3.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#performing-initial-qc"><i class="fa fa-check"></i><b>1.3.3</b> Performing initial QC</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#quality-control-for-gwas"><i class="fa fa-check"></i><b>1.4</b> Quality Control for GWAS</a><ul>
<li class="chapter" data-level="1.4.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#software-and-datasets"><i class="fa fa-check"></i><b>1.4.1</b> Software and datasets</a></li>
<li class="chapter" data-level="1.4.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#checking-the-input-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Checking the input datasets</a></li>
<li class="chapter" data-level="1.4.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#binary-ped-files"><i class="fa fa-check"></i><b>1.4.3</b> Binary PED files</a></li>
<li class="chapter" data-level="1.4.4" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#sample-qc"><i class="fa fa-check"></i><b>1.4.4</b> Sample QC:</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="github.html"><a href="github.html"><i class="fa fa-check"></i><b>2</b> GitHub</a><ul>
<li class="chapter" data-level="2.1" data-path="github.html"><a href="github.html#merge"><i class="fa fa-check"></i><b>2.1</b> Merge</a><ul>
<li class="chapter" data-level="2.1.1" data-path="github.html"><a href="github.html#troubleshooting-pages-of-interest"><i class="fa fa-check"></i><b>2.1.1</b> Troubleshooting pages of interest</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="github.html"><a href="github.html#stash"><i class="fa fa-check"></i><b>2.2</b> Stash</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="docker.html"><a href="docker.html"><i class="fa fa-check"></i><b>3</b> Docker</a><ul>
<li class="chapter" data-level="3.1" data-path="docker.html"><a href="docker.html#introduction"><i class="fa fa-check"></i><b>3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2" data-path="docker.html"><a href="docker.html#what-is-docker"><i class="fa fa-check"></i><b>3.2</b> What is docker?</a></li>
<li class="chapter" data-level="3.3" data-path="docker.html"><a href="docker.html#images-and-containers"><i class="fa fa-check"></i><b>3.3</b> Images and containers</a></li>
<li class="chapter" data-level="3.4" data-path="docker.html"><a href="docker.html#docker-hub"><i class="fa fa-check"></i><b>3.4</b> Docker Hub</a></li>
<li class="chapter" data-level="3.5" data-path="docker.html"><a href="docker.html#useful-commands"><i class="fa fa-check"></i><b>3.5</b> Useful Commands</a></li>
<li class="chapter" data-level="3.6" data-path="docker.html"><a href="docker.html#docker-coexp-web-application"><i class="fa fa-check"></i><b>3.6</b> Docker &amp; CoExp Web Application</a><ul>
<li class="chapter" data-level="3.6.1" data-path="docker.html"><a href="docker.html#software-requirements"><i class="fa fa-check"></i><b>3.6.1</b> Software requirements</a></li>
<li class="chapter" data-level="3.6.2" data-path="docker.html"><a href="docker.html#download-docker-images-of-coexp"><i class="fa fa-check"></i><b>3.6.2</b> Download Docker images of CoExp</a></li>
<li class="chapter" data-level="3.6.3" data-path="docker.html"><a href="docker.html#use-docker-composer-to-build-the-images"><i class="fa fa-check"></i><b>3.6.3</b> Use Docker-Composer to build the images</a></li>
<li class="chapter" data-level="3.6.4" data-path="docker.html"><a href="docker.html#juans-tutorial"><i class="fa fa-check"></i><b>3.6.4</b> Juan's tutorial</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html"><i class="fa fa-check"></i><b>4</b> Machine Learning Concepts</a><ul>
<li class="chapter" data-level="4.1" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#artificial-intelligence-vs.-machine-learning"><i class="fa fa-check"></i><b>4.2</b> Artificial Intelligence vs. Machine Learning</a></li>
<li class="chapter" data-level="4.3" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#supervised-learning"><i class="fa fa-check"></i><b>4.3</b> Supervised learning</a><ul>
<li class="chapter" data-level="4.3.1" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#classification"><i class="fa fa-check"></i><b>4.3.1</b> Classification</a></li>
<li class="chapter" data-level="4.3.2" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#regression"><i class="fa fa-check"></i><b>4.3.2</b> Regression</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#unsupervised-learning"><i class="fa fa-check"></i><b>4.4</b> Unsupervised learning</a><ul>
<li class="chapter" data-level="4.4.1" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#clustering"><i class="fa fa-check"></i><b>4.4.1</b> Clustering</a></li>
<li class="chapter" data-level="4.4.2" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#density-estimation"><i class="fa fa-check"></i><b>4.4.2</b> Density estimation</a></li>
<li class="chapter" data-level="4.4.3" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#visualization"><i class="fa fa-check"></i><b>4.4.3</b> Visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="aws.html"><a href="aws.html"><i class="fa fa-check"></i><b>5</b> AWS</a><ul>
<li class="chapter" data-level="5.1" data-path="aws.html"><a href="aws.html#aws-s3"><i class="fa fa-check"></i><b>5.1</b> AWS S3</a><ul>
<li class="chapter" data-level="5.1.1" data-path="aws.html"><a href="aws.html#accessing-aws-using-the-aws-command-line-interface"><i class="fa fa-check"></i><b>5.1.1</b> Accessing AWS using the AWS Command Line Interface</a></li>
<li class="chapter" data-level="5.1.2" data-path="aws.html"><a href="aws.html#checking-aws-file-integrity"><i class="fa fa-check"></i><b>5.1.2</b> Checking AWS file integrity</a></li>
<li class="chapter" data-level="5.1.3" data-path="aws.html"><a href="aws.html#creating-a-new-aws-bucket"><i class="fa fa-check"></i><b>5.1.3</b> Creating a new AWS bucket</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="aws.html"><a href="aws.html#aws-cloud-services"><i class="fa fa-check"></i><b>5.2</b> AWS cloud services</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Knowledge Book</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Knowledge Book</h1>
<p class="author"><em>Sonia García-Ruiz (<a href="mailto:s.ruiz@ucl.ac.uk">s.ruiz@ucl.ac.uk</a>)</em></p>
<p class="date" style="margin-top: 1.5em;"><em>2021-08-17</em></p>
</div>
<div id="analysis-of-genetic-association-studies" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Analysis of Genetic Association Studies</h1>
<p>The purpose of this section is to provide guidance on performing a GWAS analysis step-by step, as well as introducing to useful software for undertaking a GWAS.</p>
<div id="useful-linux-commands" class="section level2">
<h2><span class="header-section-number">1.1</span> Useful Linux commands</h2>
<p>First, I'll show you some useful Linux commands for working with GWASs files. Linux is an operating system very well suited to dealing with large datasets.</p>
<p>Create a shell script:</p>
<pre class="shell"><code>touch script.sh</code></pre>
<p>Run the script:</p>
<pre class="shell"><code>sh script.sh</code></pre>
<p>Return the number of lines of a file:</p>
<pre class="shell"><code>wc -l ./file.map</code></pre>
<p>Print only the first 15 lines of &quot;./file.map&quot; file:</p>
<pre class="shell"><code>head -15 ./file.map</code></pre>
<p>Print the bottom of the file:</p>
<pre class="shell"><code>tail -15 ./file.map</code></pre>
<p>Find the SNP &quot;rs1234&quot;:</p>
<pre class="shell"><code>grep &quot;rs1234&quot; ./file.map</code></pre>
<p>Find the SNP &quot;rs1234&quot; and print the line number of the match:</p>
<pre class="shell"><code>grep -n &quot;rs1234&quot; ./file.map</code></pre>
<p>Search the file ./file.map, for &quot;1234&quot; in the SNP name (column 2, specified by $2) and print the whole line:</p>
<pre class="shell"><code>awk &#39;$2~/1234/{print}&#39; ./file.map</code></pre>
<p>Search the file ./file.map, for &quot;1234&quot; in the SNP position (column 4, specified by $4) and print only the SNP name (column 2):</p>
<pre class="shell"><code>awk &#39;$4~/1234/{print $2}&#39; ./file.map</code></pre>
<p>Search for an exact match of &quot;rs1234676&quot; in the SNP name (column 2) and print the whole line:</p>
<pre class="shell"><code>awk &#39;$2==&quot;rs1234676&quot;/{print}&#39; ./file.map</code></pre>
<p>Search for all SNPs with a position between 90680000 up to and including 90690000 in the SNP position column (column 4) and count the lines in that output:</p>
<pre class="shell"><code>awk &#39;$4&gt;90680000&amp;&amp;$4&lt;=90690000{print}&#39; ./file.map | wc -l</code></pre>
<p>Remove the chromosome 10 from the output above:</p>
<pre class="shell"><code>awk &#39;$4&gt;90680000&amp;&amp;$4&lt;=90690000{print}&#39; ./file.map | awk &#39;$1!=10{print}&#39;</code></pre>
<p>Count the SNPs from chromosome 10 and chromosome 23:</p>
<pre class="shell"><code>awk &#39;$1==10{print}&#39; ./file.map| wc -l
awk &#39;$1==23{print}&#39; ./file.map| wc -l</code></pre>
<p>Print the SNP position of the first SNP on chromosome 23:</p>
<pre class="shell"><code>awk &#39;$1==23{print $4}&#39; ./file.map | head -1</code></pre>
<p>There are SNPs starting with other characters than &quot;rs&quot;, identify what other starting characters there are.</p>
<pre class="shell"><code>awk &#39;$2!~/rs/{print $2}&#39; ./file.map
awk &#39;$2!~/rs/&amp;&amp;$2!~/.../{print $2}&#39; ./file.map</code></pre>
</div>
<div id="useful-r-commands" class="section level2">
<h2><span class="header-section-number">1.2</span> Useful R commands</h2>
<p>NOTE: R has a system of libraries on the Comprehensive R Archive Network (CRAN) where R users deposit code to be verified and then distributed for wider use (<a href="https://cran.r-project.org/" class="uri">https://cran.r-project.org/</a>).</p>
<p>Load the file &quot;./file.map&quot; into R.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gd &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;./file.map&quot;</span>,<span class="dt">header=</span>F)
<span class="kw">names</span>(gd)&lt;-<span class="kw">c</span>(<span class="st">&quot;CHR&quot;</span>,<span class="st">&quot;SNP&quot;</span>,<span class="st">&quot;cM&quot;</span>,<span class="st">&quot;BP&quot;</span>)
<span class="kw">head</span>(gd)
<span class="kw">tail</span>(gd)</code></pre></div>
<p>Search for only rs1234568 in the SNP name (column 2):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gd[gd<span class="op">$</span>SNP<span class="op">==</span><span class="st">&quot;rs1234568&quot;</span>,]</code></pre></div>
<p>Search for all SNPs with a position between 90680000 up to and including 90690000 in the SNP position column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gd[gd<span class="op">$</span>BP<span class="op">&gt;</span><span class="dv">90680000</span><span class="op">&amp;</span>gd<span class="op">$</span>BP<span class="op">&lt;=</span><span class="dv">90690000</span>,]</code></pre></div>
<p>Do you have any genome-wide significant (p=5x10-8) SNPs?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gd[gd<span class="op">$</span>P<span class="op">&lt;</span><span class="fl">0.00000005</span>,]</code></pre></div>
<p>Report results using a Manhattan plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(gd<span class="op">$</span>BP,<span class="op">-</span><span class="kw">log</span>(gd<span class="op">$</span>P,<span class="dt">base=</span><span class="dv">10</span>))</code></pre></div>
<p>There were SNPs from both chromosome 10 and 23 in the .map file. Separate chromosomes by colour:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(gd<span class="op">$</span>BP,<span class="op">-</span><span class="kw">log</span>(gd<span class="op">$</span>P,<span class="dt">base=</span><span class="dv">10</span>), <span class="dt">col=</span>gd<span class="op">$</span>CHR)</code></pre></div>
<p>Add a reference line to know which SNPs have a p-value below 1x10-5:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(gd<span class="op">$</span>BP,<span class="op">-</span><span class="kw">log</span>(gd<span class="op">$</span>P,<span class="dt">base=</span><span class="dv">10</span>), <span class="dt">col=</span>gd<span class="op">$</span>CHR)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="op">-</span><span class="kw">log</span>(<span class="fl">1e-5</span>,<span class="dt">base=</span><span class="dv">10</span>),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p>Use the package &quot;qqman&quot; to generate a Manhattan plot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">manhattan</span>(gd)
<span class="kw">manhattan</span>(gd[gd<span class="op">$</span>chromosome<span class="op">==</span><span class="dv">10</span>,], <span class="dt">chr=</span><span class="st">&quot;chromosome&quot;</span>,<span class="dt">bp=</span><span class="st">&quot;position&quot;</span>,<span class="dt">snp=</span><span class="st">&quot;rsid&quot;</span>,<span class="dt">p=</span><span class="st">&quot;pvalue&quot;</span>)
<span class="kw">manhattan</span>(gd[gd<span class="op">$</span>chromosome<span class="op">==</span><span class="dv">23</span>,], <span class="dt">chr=</span><span class="st">&quot;chromosome&quot;</span>,<span class="dt">bp=</span><span class="st">&quot;position&quot;</span>,<span class="dt">snp=</span><span class="st">&quot;rsid&quot;</span>,<span class="dt">p=</span><span class="st">&quot;pvalue&quot;</span>)</code></pre></div>
</div>
<div id="variant-calling-from-sequencing-data" class="section level2">
<h2><span class="header-section-number">1.3</span> Variant Calling from sequencing data</h2>
<div id="generate-genotype-calls-for-a-single-sample" class="section level3">
<h3><span class="header-section-number">1.3.1</span> Generate genotype calls for a single sample</h3>
<p>In this section we will generate genotype calls from aligned sequencing data, explore the calls and perform some initial quality checks.</p>
<p>A commonly used variant caller comes from the Genome Analysis Toolkit (GATK).</p>
<ul>
<li>GATK takes files containing aligned sequencing reads. These are called .bam files.</li>
<li>GATK produces .vcf files containing details of the variants, to explore and manipulate these files, we'll be using vcftools. This is also freely available and can be found: <a href="http://vcftools.sourceforge.net/" class="uri">http://vcftools.sourceforge.net/</a></li>
<li>To visualise the variant calls we'll be using Integrated Genome Viewer (IGV); this is available from: <a href="http://software.broadinstitute.org/software/igv/" class="uri">http://software.broadinstitute.org/software/igv/</a></li>
</ul>
<p>Usually, each .bam file has an accompanying .bai file. These are index files, which are programs used to navigate the data faster – they must be present for software programs to run.</p>
<p>Generate genotype calls for the sample NA12878 using GATK:</p>
<pre class="shell"><code>gatk -T HaplotypeCaller \
-R human_g1k_b37_20.fasta \
-I NA12878_wgs_20.bam \
-o yourfilename.vcf \
-L 20:10,000,000-10,200,000</code></pre>
<p>Variant calling on a whole genome can take a long time, we're only looking at a very small region: Chromosome 20, positions 10,000,000 – 10,200,000, specified by the –L option.</p>
<p>(*) Details of the different options used: <a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php" class="uri">https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php</a></p>
<p>To have a look at the first 5 variant calls using:</p>
<pre class="shell"><code>head -34 yourfilename.vcf | tail -6</code></pre>
<p>Filter the vcf file and see only the variants called in the region 20:10,002,371-10,002,546:</p>
<pre class="shell"><code>vcftools --vcf yourfilename.vcf \
--chr 20 \
--from-bp 10002371 \
--to-bp 10002546 \
--out yournewfilename \
--recode</code></pre>
<p>The code above will generate the files: yournewfilename.recode.vcf and yournewfilename.log.</p>
<p>To know how many variants were identified in this region:</p>
<pre class="shell"><code>cat yournewfilename.log</code></pre>
<p>What are those variants?</p>
<pre class="shell"><code>cat yournewfilename.recode.vcf</code></pre>
</div>
<div id="generating-genotypes-from-multiple-.bam-files" class="section level3">
<h3><span class="header-section-number">1.3.2</span> Generating genotypes from multiple .bam files</h3>
<p>To generate genotypes for multiple samples, one option is to run the same command than above but give multiple -I options:</p>
<pre class="shell"><code>gatk -T HaplotypeCaller \
-R human_g1k_b37_20.fasta \
-I NA12878_wgs_20.bam \
-I NA12882_wgs_20.bam \
-I NA12877_wgs_20.bam \
-o differentfilename.vcf \
-L 20:10,000,000-10,200,000</code></pre>
<p>Another option is to run each sample independently, but generate a .g.vcf file and then merge the .g.vcf files to get a single call set.</p>
<p>Generate a .g.vcf file for sample NA12878 (then run the same command on the bam files for samples NA12877 and NA12882):</p>
<pre class="shell"><code>gatk -T HaplotypeCaller \
-R human_g1k_b37_20.fasta \
-I NA12878_wgs_20.bam \
-o yourchoice_NA12878.g.vcf \
-ERC GVCF \
-L 20:10,000,000-10,200,000</code></pre>
<p>The main difference between the .g.vcf and the .vcf file is that the gvcf file contains a line per base or region if the region contains no variants. The gvcf also includes <NON_REF> as a symbolic allele, this allows for complex variants at the site and to makes it possible to give a confidence measure in the reference allele (more info: <a href="http://gatkforums.broadinstitute.org/gatk/discussion/4017/what-is-a-gvcf-and-how-is-it-differentfrom-a-regular-vcf" class="uri">http://gatkforums.broadinstitute.org/gatk/discussion/4017/what-is-a-gvcf-and-how-is-it-differentfrom-a-regular-vcf</a>)</p>
<p>Merge the individual .g.vcf files and get a .vcf file containing the calls for all samples:</p>
<pre class="shell"><code>gatk -T GenotypeGVCFs \
-R human_g1k_b37_20.fasta \
-V yourchoice_NA12878.g.vcf \
-V yourchoice_NA12882.g.vcf \
-V yourchoice_NA12877.g.vcf \
-o yourchoice_gvcf_jointcalls.vcf \
-L 20:10,000,000-10,200,000</code></pre>
</div>
<div id="performing-initial-qc" class="section level3">
<h3><span class="header-section-number">1.3.3</span> Performing initial QC</h3>
<p>GWAS studies often focus only on SNPS, whereas variant callers identify more variants, such as indels. To create a file containing just SNPs (no indels or other variants) the --remove-indels option from vcftools can be used:</p>
<pre class="shell"><code>vcftools --vcf yourchoice_gvcf_jointcalls.vcf \
--remove-indels \
--out yourchoice_snps \
--recode</code></pre>
<p>Calculate the Ts/Tv ratio:</p>
<pre class="shell"><code>vcftools --vcf yourchoice_gvcf_jointcalls.vcf \
--TsTv-summary \
--out yourchoice_tstv</code></pre>
<p>The QualByDepth (QD) score is the variant quality (QUAL field) divided by the unfiltered depth of nonreference samples. Variants with a low QD may be unreliable and have a high false positive rate. To add a filter to the .vcf file which flags variants which have a quality-by-depth less than 2.0:</p>
<pre class="shell"><code>gatk -T VariantFiltration \
-R human_g1k_b37_20.fasta \
-V yourchoice_gvcf_jointcalls.vcf \
--filterExpression &#39;QD&lt;2.0&#39; \
--filterName &#39;QualityDepth&#39; \
-o yourchoice_flagged.vcf</code></pre>
<p>NOTE: SNPs and Indels may need to be treated differently – to do this, the file can be first splitted into 2 vcfs, one containing SNPs and one containing indels, then add the flags should be added.</p>
<p>The new .vcf is now different from the original as the FILTER column is now filled in, usually with PASS.</p>
<p>Remove all SNPs with a filter flag other than PASS:</p>
<pre class="shell"><code>vcftools --vcf yourchoice_flagged.vcf \
--remove-filtered-all \
--recode \
--out yourchoice_filtered</code></pre>
<p>To convert the SNP filtered .vcf file to PLINK .ped and .map files:</p>
<pre class="shell"><code>vcftools --vcf yourchoice_gvcf_jointcalls.vcf \
--remove-indels \
--out yourchoice_plink</code></pre>
</div>
</div>
<div id="quality-control-for-gwas" class="section level2">
<h2><span class="header-section-number">1.4</span> Quality Control for GWAS</h2>
<div id="software-and-datasets" class="section level3">
<h3><span class="header-section-number">1.4.1</span> Software and datasets</h3>
<p>A commonly used program for GWAS analysis is called 'PLINK', which is freely available for download. The webpage for PLINK is available at: <a href="http://pngu.mgh.harvard.edu/~purcell/plink/" class="uri">http://pngu.mgh.harvard.edu/~purcell/plink/</a> For undertaking a GWAS, two input files are required: one '.ped' file and one '.map' file.</p>
<ul>
<li>The PED file is a white-space (space or tab) delimited file. It includes one row for each participant within the study, and at least seven columns. Each of the SNPs genotyped is represented by two columns (column 7 onwards) –one for each of the two alleles held by the individual at the SNP. Typical coding would be A, C, G or T to represent the four possible alleles, or similarly 1, 2, 3 or 4. Coding for missing genotype is '0 0'.</li>
<li>The .map file includes information on the SNPs genotyped, and each row represents a SNP. It includes 4 columns.</li>
</ul>
<p>In this example, we will be working on analysing the GWAS dataset 'genstudy' (a .map and .ped file for 'genstudy'). The study is a case-control study with 252 diseased (cases) and 252 non-diseased (controls) individuals. All individuals have been genotyped for 198,684 SNPs on chromosome 10 and 22,379 SNPs on chromosome X. The purpose of the study is to try and identify SNPs on chromosome 10 associated with disease.</p>
<p>To view the first row and first eight columns of the 'genstudy.ped' file:</p>
<pre class="shell"><code>awk &#39;FNR == 1 { print $1,$2,$3,$4,$5,$6,$7,$8} &#39; genstudy.ped</code></pre>
<p>The eight columns correspond to the first six mandatory columns plus two columns representing the first SNP genotyped.</p>
</div>
<div id="checking-the-input-datasets" class="section level3">
<h3><span class="header-section-number">1.4.2</span> Checking the input datasets</h3>
<p>Before starting a genotype QC or GWAS association analysis, it is important to check that the input files are in good order, and contain the individuals and SNPs that we think they should. Do this by focussing on the SNPs on chromosome 10:</p>
<pre class="shell"><code>plink --file genstudy --chr 10 --noweb</code></pre>
<p>Note: the '-noweb' option is included to make PLINK run faster – otherwise it connects to the web to check for updates before running the analysis.</p>
</div>
<div id="binary-ped-files" class="section level3">
<h3><span class="header-section-number">1.4.3</span> Binary PED files</h3>
<p>To save space and time, .map and .ped files can be converted to binary format.</p>
<p>To create a set of binary format files for the 'genstudy' dataset:</p>
<pre class="shell"><code>plink --file genstudy --make-bed --out genstudy --noweb</code></pre>
<p>To use the binary format files instead of the .ped and .map files, we just substitute the option --file with --bfile in the PLINK command line:</p>
<pre class="shell"><code>plink --file genstudy --chr 10 --noweb</code></pre>
</div>
<div id="sample-qc" class="section level3">
<h3><span class="header-section-number">1.4.4</span> Sample QC:</h3>
<p>In this section, we will start undertaking genotype QC on the 'genstudy' dataset – starting first of all with sample QC, one step at a time.</p>
<div id="checks-for-missing-data" class="section level4">
<h4><span class="header-section-number">1.4.4.1</span> Checks for missing data</h4>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="github.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["knowledge-book.pdf", "knowledge-book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
