<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 eQTL | Knowledge Book</title>
  <meta name="description" content="This book contains some of the knowledge I have been acquiring during my PhD studies." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 eQTL | Knowledge Book" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book contains some of the knowledge I have been acquiring during my PhD studies." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 eQTL | Knowledge Book" />
  
  <meta name="twitter:description" content="This book contains some of the knowledge I have been acquiring during my PhD studies." />
  

<meta name="author" content="Sonia GarcÃ­a-Ruiz (s.ruiz@ucl.ac.uk)" />


<meta name="date" content="2022-03-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="analysis-of-genetic-association-studies.html"/>
<link rel="next" href="aws.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Sonia's Knowledge Book</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html"><i class="fa fa-check"></i><b>1</b> Analysis of Genetic Association Studies</a><ul>
<li class="chapter" data-level="1.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#useful-linux-commands"><i class="fa fa-check"></i><b>1.1</b> Useful Linux commands</a></li>
<li class="chapter" data-level="1.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#useful-r-commands"><i class="fa fa-check"></i><b>1.2</b> Useful R commands</a></li>
<li class="chapter" data-level="1.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#variant-calling-from-sequencing-data"><i class="fa fa-check"></i><b>1.3</b> Variant Calling from sequencing data</a><ul>
<li class="chapter" data-level="1.3.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#generate-genotype-calls-for-a-single-sample"><i class="fa fa-check"></i><b>1.3.1</b> Generate genotype calls for a single sample</a></li>
<li class="chapter" data-level="1.3.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#generating-genotypes-from-multiple-.bam-files"><i class="fa fa-check"></i><b>1.3.2</b> Generating genotypes from multiple .bam files</a></li>
<li class="chapter" data-level="1.3.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#performing-initial-qc"><i class="fa fa-check"></i><b>1.3.3</b> Performing initial QC</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#quality-control-for-gwas"><i class="fa fa-check"></i><b>1.4</b> Quality Control for GWAS</a><ul>
<li class="chapter" data-level="1.4.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#software-and-datasets"><i class="fa fa-check"></i><b>1.4.1</b> Software and datasets</a></li>
<li class="chapter" data-level="1.4.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#checking-the-input-datasets"><i class="fa fa-check"></i><b>1.4.2</b> Checking the input datasets</a></li>
<li class="chapter" data-level="1.4.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#binary-ped-files"><i class="fa fa-check"></i><b>1.4.3</b> Binary PED files</a></li>
<li class="chapter" data-level="1.4.4" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#sample-qc"><i class="fa fa-check"></i><b>1.4.4</b> Sample QC</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#testing-for-associations"><i class="fa fa-check"></i><b>1.5</b> Testing for Associations</a><ul>
<li class="chapter" data-level="1.5.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#software-and-datasets-1"><i class="fa fa-check"></i><b>1.5.1</b> Software and datasets</a></li>
<li class="chapter" data-level="1.5.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#univariate-analyses-of-association"><i class="fa fa-check"></i><b>1.5.2</b> Univariate analyses of association</a></li>
<li class="chapter" data-level="1.5.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#multiple-regression-analyses"><i class="fa fa-check"></i><b>1.5.3</b> Multiple regression analyses</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#population-structure-in-gwas"><i class="fa fa-check"></i><b>1.6</b> Population structure in GWAS</a><ul>
<li class="chapter" data-level="1.6.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#testing-for-association-under-an-additive-model"><i class="fa fa-check"></i><b>1.6.1</b> Testing for association under an additive model</a></li>
<li class="chapter" data-level="1.6.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#performing-multi-dimensional-scaling"><i class="fa fa-check"></i><b>1.6.2</b> Performing multi-dimensional scaling</a></li>
<li class="chapter" data-level="1.6.3" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#testing-for-association-under-an-additive-model-with-adjustment-for-confounding"><i class="fa fa-check"></i><b>1.6.3</b> Testing for association under an additive model with adjustment for confounding</a></li>
</ul></li>
<li class="chapter" data-level="1.7" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#phasing-and-imputation"><i class="fa fa-check"></i><b>1.7</b> Phasing and Imputation</a><ul>
<li class="chapter" data-level="1.7.1" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#pre-phasing-and-imputation"><i class="fa fa-check"></i><b>1.7.1</b> Pre-phasing and imputation</a></li>
<li class="chapter" data-level="1.7.2" data-path="analysis-of-genetic-association-studies.html"><a href="analysis-of-genetic-association-studies.html#re-running-our-univariate-analyses-of-association"><i class="fa fa-check"></i><b>1.7.2</b> Re-running our univariate analyses of association</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="eqtl.html"><a href="eqtl.html"><i class="fa fa-check"></i><b>2</b> eQTL</a><ul>
<li class="chapter" data-level="2.1" data-path="eqtl.html"><a href="eqtl.html#maf"><i class="fa fa-check"></i><b>2.1</b> MAF</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="aws.html"><a href="aws.html"><i class="fa fa-check"></i><b>3</b> AWS</a><ul>
<li class="chapter" data-level="3.1" data-path="aws.html"><a href="aws.html#aws-s3"><i class="fa fa-check"></i><b>3.1</b> AWS S3</a><ul>
<li class="chapter" data-level="3.1.1" data-path="aws.html"><a href="aws.html#configure-the-aws-cli-service"><i class="fa fa-check"></i><b>3.1.1</b> Configure the AWS CLI service</a></li>
<li class="chapter" data-level="3.1.2" data-path="aws.html"><a href="aws.html#upload-a-single-file-to-aws-using-aws-cli"><i class="fa fa-check"></i><b>3.1.2</b> Upload a single file to AWS using AWS CLI</a></li>
<li class="chapter" data-level="3.1.3" data-path="aws.html"><a href="aws.html#upload-multiple-files-to-aws-using-aws-cli"><i class="fa fa-check"></i><b>3.1.3</b> Upload multiple files to AWS using AWS CLI</a></li>
<li class="chapter" data-level="3.1.4" data-path="aws.html"><a href="aws.html#download-a-single-file-from-aws-using-aws-cli"><i class="fa fa-check"></i><b>3.1.4</b> Download a single file from AWS using AWS CLI</a></li>
<li class="chapter" data-level="3.1.5" data-path="aws.html"><a href="aws.html#download-multiple-files-from-aws-using-aws-cli"><i class="fa fa-check"></i><b>3.1.5</b> Download multiple files from AWS using AWS CLI</a></li>
<li class="chapter" data-level="3.1.6" data-path="aws.html"><a href="aws.html#requester-payer"><i class="fa fa-check"></i><b>3.1.6</b> Requester payer</a></li>
<li class="chapter" data-level="3.1.7" data-path="aws.html"><a href="aws.html#checking-aws-file-integrity"><i class="fa fa-check"></i><b>3.1.7</b> Checking AWS file integrity</a></li>
<li class="chapter" data-level="3.1.8" data-path="aws.html"><a href="aws.html#creating-a-new-aws-bucket"><i class="fa fa-check"></i><b>3.1.8</b> Creating a new AWS bucket</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="aws.html"><a href="aws.html#aws-ec2"><i class="fa fa-check"></i><b>3.2</b> AWS EC2</a><ul>
<li class="chapter" data-level="3.2.1" data-path="aws.html"><a href="aws.html#generating-an-ec2-instance"><i class="fa fa-check"></i><b>3.2.1</b> Generating an EC2 instance</a></li>
<li class="chapter" data-level="3.2.2" data-path="aws.html"><a href="aws.html#connecting-to-an-ec2-instance"><i class="fa fa-check"></i><b>3.2.2</b> Connecting to an EC2 instance</a></li>
<li class="chapter" data-level="3.2.3" data-path="aws.html"><a href="aws.html#installingrunning-rstudio-on-an-ec2"><i class="fa fa-check"></i><b>3.2.3</b> Installing/Running RStudio on an EC2</a></li>
<li class="chapter" data-level="3.2.4" data-path="aws.html"><a href="aws.html#access-rstudio-using-your-browser"><i class="fa fa-check"></i><b>3.2.4</b> Access RStudio using your browser</a></li>
<li class="chapter" data-level="3.2.5" data-path="aws.html"><a href="aws.html#configuring-https-with-rstudio-using-aws-ec2"><i class="fa fa-check"></i><b>3.2.5</b> Configuring HTTPS with RStudio using AWS EC2</a></li>
<li class="chapter" data-level="3.2.6" data-path="aws.html"><a href="aws.html#how-to-transfer-files-from-ec2-to-s3"><i class="fa fa-check"></i><b>3.2.6</b> How to transfer files from EC2 to S3</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="aws.html"><a href="aws.html#aws-list-of-cloud-services"><i class="fa fa-check"></i><b>3.3</b> AWS List of Cloud Services</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="docker.html"><a href="docker.html"><i class="fa fa-check"></i><b>4</b> Docker</a><ul>
<li class="chapter" data-level="4.1" data-path="docker.html"><a href="docker.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="docker.html"><a href="docker.html#what-is-docker"><i class="fa fa-check"></i><b>4.2</b> What is docker?</a></li>
<li class="chapter" data-level="4.3" data-path="docker.html"><a href="docker.html#images-and-containers"><i class="fa fa-check"></i><b>4.3</b> Images and containers</a></li>
<li class="chapter" data-level="4.4" data-path="docker.html"><a href="docker.html#docker-hub"><i class="fa fa-check"></i><b>4.4</b> Docker Hub</a></li>
<li class="chapter" data-level="4.5" data-path="docker.html"><a href="docker.html#useful-commands"><i class="fa fa-check"></i><b>4.5</b> Useful Commands</a></li>
<li class="chapter" data-level="4.6" data-path="docker.html"><a href="docker.html#docker-coexp-web-application"><i class="fa fa-check"></i><b>4.6</b> Docker &amp; CoExp Web Application</a><ul>
<li class="chapter" data-level="4.6.1" data-path="docker.html"><a href="docker.html#software-requirements"><i class="fa fa-check"></i><b>4.6.1</b> Software requirements</a></li>
<li class="chapter" data-level="4.6.2" data-path="docker.html"><a href="docker.html#download-docker-images-of-coexp"><i class="fa fa-check"></i><b>4.6.2</b> Download Docker images of CoExp</a></li>
<li class="chapter" data-level="4.6.3" data-path="docker.html"><a href="docker.html#use-docker-composer-to-build-the-images"><i class="fa fa-check"></i><b>4.6.3</b> Use Docker-Composer to build the images</a></li>
<li class="chapter" data-level="4.6.4" data-path="docker.html"><a href="docker.html#juans-tutorial"><i class="fa fa-check"></i><b>4.6.4</b> Juan's tutorial</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="github.html"><a href="github.html"><i class="fa fa-check"></i><b>5</b> GitHub</a><ul>
<li class="chapter" data-level="5.1" data-path="github.html"><a href="github.html#merge"><i class="fa fa-check"></i><b>5.1</b> Merge</a><ul>
<li class="chapter" data-level="5.1.1" data-path="github.html"><a href="github.html#troubleshooting-pages-of-interest"><i class="fa fa-check"></i><b>5.1.1</b> Troubleshooting pages of interest</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="github.html"><a href="github.html#stash"><i class="fa fa-check"></i><b>5.2</b> Stash</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="globus.html"><a href="globus.html"><i class="fa fa-check"></i><b>6</b> Globus</a><ul>
<li class="chapter" data-level="6.1" data-path="globus.html"><a href="globus.html#useful-commands-1"><i class="fa fa-check"></i><b>6.1</b> Useful Commands</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html"><i class="fa fa-check"></i><b>7</b> Machine Learning Concepts</a><ul>
<li class="chapter" data-level="7.1" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#introduction-1"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#artificial-intelligence-vs.-machine-learning"><i class="fa fa-check"></i><b>7.2</b> Artificial Intelligence vs. Machine Learning</a></li>
<li class="chapter" data-level="7.3" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#supervised-learning"><i class="fa fa-check"></i><b>7.3</b> Supervised learning</a><ul>
<li class="chapter" data-level="7.3.1" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#classification"><i class="fa fa-check"></i><b>7.3.1</b> Classification</a></li>
<li class="chapter" data-level="7.3.2" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#regression"><i class="fa fa-check"></i><b>7.3.2</b> Regression</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#unsupervised-learning"><i class="fa fa-check"></i><b>7.4</b> Unsupervised learning</a><ul>
<li class="chapter" data-level="7.4.1" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#clustering"><i class="fa fa-check"></i><b>7.4.1</b> Clustering</a></li>
<li class="chapter" data-level="7.4.2" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#density-estimation"><i class="fa fa-check"></i><b>7.4.2</b> Density estimation</a></li>
<li class="chapter" data-level="7.4.3" data-path="machine-learning-concepts.html"><a href="machine-learning-concepts.html#visualization"><i class="fa fa-check"></i><b>7.4.3</b> Visualization</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Knowledge Book</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="eqtl" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> eQTL</h1>
<p>Adapted from <a href="https://www.um.edu.mt/__data/assets/pdf_file/0011/289487/eQTL_praticals_final.pdf">eQTL_praticals_final.</a>.</p>
<div id="maf" class="section level2">
<h2><span class="header-section-number">2.1</span> MAF</h2>
<p>MAF stands for Minor allele frequency. Every person has approximately 4-5 million bases that differ from the reference genome. Some of those variants are common in a certain population, others not. The MAF measure calculates the rareness of a specific variant (allele).</p>
<p>In one genomic position, there can be up to four different alleles (A, C, G, T). When an allele differs from the reference genome, it is called single nucleotide polymorphism. They can also be homozygous (maternal and paternal alleles are similar) or heterozygous (when the maternal allele is different from the paternal allele at that position).</p>
<p>SNPs in a population are always defined by the genomic position and by two alleles: The allele defined in the reference genome and one allele present in some individuals, but different from the reference sequence.</p>
<p>In eQTL analyses a SNP is defined as a single allele from an individual sample that has been found to differ from the reference genome. In case different samples present multiple different alleles for one position, those alleles will be treated as independent SNPs. In other words, SNPs in eQTL analyses can only have two alleles: the allele from the reference genome and the allele found in the sample.</p>
<p>The MAF measure shows the percentage of alleles within a population. At an individual level, each sample can have 0, 1 or 2 times the alternative (i.e. the one different from the reference) allele.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="analysis-of-genetic-association-studies.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="aws.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["knowledge-book.pdf", "knowledge-book.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
